runtime/compiler/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,245 @@
+      TR::Block *originalBlock = testLocationIter->first.second;
+      TR::Node *node = asyncCheckTreeTop->getNode();
+      int32_t depth = testLocationIter->second;
+      if (trace())
+         traceMsg(comp, "block_%d, n%dn, depth = %d\n",originalBlock->getNumber(), asyncCheckTreeTop->getNode()->getGlobalIndex(), depth);
+      TR_ByteCodeInfo bci = asyncCheckTreeTop->getNode()->getByteCodeInfo();
+      
+      TR::Node *root = bfi->generateBlockRawCountCalculationSubTree(comp, node, trace());
+      
+      // If we got a bad counters/ bad block frequency info, above API would return NULL, if that is the case we can not generate a recompilation test for that location.
+      if (!root)
+         {
+         TR::DebugCounter::incStaticDebugCounter(comp, TR::DebugCounter::debugCounterName(comp, "jprofiling.instrument/badcounters/(%s)", comp->signature()));
+         continue;
+         }
+      dumpOptDetails(comp, "%s Add recompilation test after asyncCheck node n%dn\n", optDetailString(), node->getGlobalIndex());
+
+      //Splitting the original block from AsyncCheckTT
+      TR::Block *remainingCodeBlock = originalBlock->split(asyncCheckTreeTop->getNextTreeTop(), cfg, true, true);
+      TR::Block *callRecompileBlock = TR::Block::createEmptyBlock(node, comp, UNKNOWN_COLD_BLOCK_COUNT);
+      callRecompileBlock->setIsCold(true);
+      
+      // jitRetranslateCallerWithPrep Helper call
+      TR::TreeTop *callTree = TR::TransformUtil::generateRetranslateCallerWithPrepTrees(node, TR_PersistentMethodInfo::RecompDueToJProfiling, comp);
+      callTree->getNode()->setIsProfilingCode();
+      callRecompileBlock->append(callTree);
+      cfg->addNode(callRecompileBlock);
+      const char * const name = TR::DebugCounter::debugCounterName(comp,
+      "recompilationHelper/(%s)/%d",
+      comp->signature(),depth);
+      TR::DebugCounter::prependDebugCounter(comp, name, callTree);
+      
+      // Code to calculate the raw frequency of loop from block counters and comparing with the adjusted recompilation threshold.
+      // threshold for this particular test location is calculated from the base recompile threshold and the nesting depth of the loop
+      // Putting a higher threshold limit to 10K currently to prohibit running profiling body for too long.
+      int32_t threshold = recompileThreshold << ((depth-1)*1);
+      TR::Node *cmpNode = TR::Node::createif(TR::ificmple, root, TR::Node::iconst(node, threshold <= maxLoopRecompilationThreshold ? threshold : maxLoopRecompilationThreshold), remainingCodeBlock->getEntry());
+      TR::TreeTop *cmpFlag = TR::TreeTop::create(comp, cmpNode);
+      cmpFlag->getNode()->setIsProfilingCode();
+      remainingCodeBlock->getEntry()->insertTreeTopsBeforeMe(callRecompileBlock->getEntry(), callRecompileBlock->getExit());
+      static bool dontGenerateJProfilingRecompQueueTest = (feGetEnv("TR_DontGenerateJProfilingRecompQueueTest") != NULL);
+      if (!dontGenerateJProfilingRecompQueueTest)
```

Comment:
Hmm this reads badly - not dontGenerate... Can you get rid of one of these negatives - it requires too much thinking to figure out what is going on...

Reply:
Changed name: https://github.com/eclipse/openj9/pull/2348/commits/62454581a8c7a177249a55fcecccd3b9a5b793fa